### PR TITLE
refactor: Migrates OnlineArchive resource to new Atlas SDK

### PR DESCRIPTION
**Pull request by lantoli**
## Proposed changes

_Jira ticket:_ [INTMDB-1097](https://jira.mongodb.org/browse/INTMDB-1097)

Migrating OnlineArchive resource to new Atlas SDK.

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
  expected)
- [ ] This change requires a documentation update

## Manual QA performed:

- [ ] cfn invoke for each of CRUDL/cfn test
- [ ] Updated resource in  [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples)
- [x] Published to AWS private registry
- [x] Used the template in [example](https://github.com/mongodb/mongodbatlas-cloudformation-resources/tree/master/examples) to create and update a stack in AWS
- [x] Deleted stack to ensure resources are deleted
- [ ] Created multiple resources in same stack
- [x] Validated in Atlas UI
- [x] Included screenshots

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added any necessary documentation (if appropriate)
- [x] I have run `make fmt` and formatted my code
- [ ] For CFN Resources: I have released by changes in the private registry and proved by change
  works in Atlas

## Further comments

**Contract testing:**
contract testing

**Manual testing creating the stack using private registry:**
create cf
create atlas

**Manual testing deleting the stack using private registry:**
delete cf
delete atlas

### CODE REVIEW

File: cfn-resources/online-archive/Makefile [hunk 1]
```diff
@@ -1,11 +1,11 @@
-.PHONY: build test clean
+.PHONY: build debug clean create-test-resources delete-test-resources run-contract-testing
```
[the remark this answers is not in the document]
Comment (lantoli) [reply cfn-resources/online-archive/Makefile:1]: no changing, I'm fixing the PHONY list of actions to match the Makefile file, e.g. test doesn't exist

File: cfn-resources/online-archive/Makefile [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
 tags=logging callback metrics scheduler
 cgo=0
 goos=linux
 goarch=amd64
 CFNREP_GIT_SHA?=$(shell git rev-parse HEAD)
 ldXflags=-s -w -X github.com/mongodb/mongodbatlas-cloudformation-resources/util.defaultLogLevel=info -X github.com/mongodb/mongodbatlas-cloudformation-resources/version.Version=${CFNREP_GIT_SHA}
-ldXflagsD=-s -w -X github.com/mongodb/mongodbatlas-cloudformation-resources/util.defaultLogLevel=debug -X github.com/mongodb/mongodbatlas-cloudformation-resources/version.Version=${CFNREP_GIT_SHA}
+ldXflagsD=-X github.com/mongodb/mongodbatlas-cloudformation-resources/util.defaultLogLevel=debug -X github.com/mongodb/mongodbatlas-cloudformation-resources/version.Version=${CFNREP_GIT_SHA}
```
[the remark this answers is not in the document]
Comment (lantoli) [reply cfn-resources/online-archive/Makefile:8]: ldXflagsD is for debug ldXflags is for make build so it's not affected with this change

File: cfn-resources/online-archive/cmd/resource/resource.go [hunk 3]
```diff
@@ -47,24 +46,17 @@ func setup() {
...
-	modelValidation := validateModel(CreateRequiredFields, currentModel)
-	if modelValidation != nil {
-		return *modelValidation, nil
+	if err := validator.ValidateModel(CreateRequiredFields, currentModel); err != nil {
+		return *err, nil
 	}
 
 	if currentModel.Profile == nil || *currentModel.Profile == "" {
 		currentModel.Profile = aws.String(profile.DefaultProfile)
 	}
```
[the remark this answers is not in the document]
Comment (lantoli) [reply cfn-resources/online-archive/cmd/resource/resource.go:57]: ha ha, will do :-)

File: cfn-resources/online-archive/cmd/resource/resource.go [hunk 4]
```diff
@@ -76,19 +68,17 @@ func Create(req handler.Request, prevModel *Model, currentModel *Model) (handler
 		return validateProgress(ctx, client, currentModel, "PENDING")
 	}
 
-	inputRequest, errHandler := mapToArchivePayload(currentModel)
+	inputRequest, errHandler := prepareCreate(currentModel)
```
[the remark this answers is not in the document]
Comment (lantoli) [reply cfn-resources/online-archive/cmd/resource/resource.go:71]: this is to prepare the params to Create, maybe newCreateParams ?

File: cfn-resources/online-archive/cmd/resource/resource.go [hunk 5]
```diff
@@ -43,31 +42,22 @@ var requiredCriteriaType = map[string][]string{
 	"CUSTOM": {"Query"},
 }
 
-func setup() {
+func setupLogAndProfile(currentModel *Model) {
```
Comment (marcosuma) [inline cfn-resources/online-archive/cmd/resource/resource.go:45]: nit: Could we have a `setupLogs` and `setupProfile`? Method names with `And` inside smell like an aggregation of responsibility that will likely increase in the future.

Comment (lantoli) [reply cfn-resources/online-archive/cmd/resource/resource.go:45]: I see your point but they're really very small methods, almost one line each.  What about keeping the name "setup", at the end the responsibility is to setup the methods, I wanted to be more explicit about what is set up but maybe it's not needed.

Comment (marcosuma) [reply cfn-resources/online-archive/cmd/resource/resource.go:45]: "setup()" methods are so generic that will basically always force you to go look inside what is doing.

Understand your concern of "it's just one line", but I guess the reason why we are aggregating this into separate funcs is to remove boilerplate code repeated across the four methods and make it easy to read, is that correct? 

I'll leave that up to you, if you think that the option of having separate dedicated methods won't be useful then I don't have strong opinions.

Comment (AgustinBettati) [reply cfn-resources/online-archive/cmd/resource/resource.go:45]: I like the idea of having a separate```setDefaultProfileIfNotDefined```, and this function can be defined in a common package as all resources will make use of the same implementation.

File: cfn-resources/online-archive/cmd/resource/resource.go [hunk 6]
```diff
@@ -272,17 +178,15 @@ func Delete(req handler.Request, prevModel *Model, currentModel *Model) (handler
 		return validateProgress(ctx, client, currentModel, "PENDING")
 	}
 
-	a, _ := ArchiveExists(context.Background(), client, currentModel)
-	if a.State == "DELETED" {
+	if a, _ := ArchiveExists(ctx, client, currentModel); *a.State == "DELETED" {
```
Comment (AgustinBettati) [inline cfn-resources/online-archive/cmd/resource/resource.go:181]:
```suggestion
        deleted := "DELETED" 
	if a, _ := ArchiveExists(ctx, client, currentModel); util.AreStringPtrEqual(a.State, &deleted) {
```
Small suggestions for avoiding nil dereferencing. We can create a variant of this method to receive a string value as well.

File: cfn-resources/online-archive/cmd/resource/resource.go (hunk 5 above)
Comment (lantoli) [reply cfn-resources/online-archive/cmd/resource/resource.go:45]: I like that

File: cfn-resources/online-archive/cmd/resource/resource.go [hunk 7]
```diff
@@ -361,18 +330,25 @@ func validateProgress(ctx context.Context, client *mongodbatlas.Client, currentM
 	p := handler.NewProgressEvent()
 	p.OperationStatus = cloudformation.OperationStatusSuccess
 	p.Message = "Complete"
-	if archive.State != "DELETED" {
+	if *archive.State != "DELETED" {
```
Comment (andreaangiolillo) [inline cfn-resources/online-archive/cmd/resource/resource.go:333]: should DELETE be a constant?

**Review by andreaangiolillo — DISMISSED** [review]
LGTM

File: cfn-resources/online-archive/cmd/resource/resource.go (hunk 7 above)
Comment (lantoli) [reply cfn-resources/online-archive/cmd/resource/resource.go:334]: i changed to be used only one place so i suppose it's not some important now

**Review by andreaangiolillo — APPROVED** [review]

File: cfn-resources/online-archive/cmd/resource/resource.go (hunk 6 above)
Comment (lantoli) [reply cfn-resources/online-archive/cmd/resource/resource.go]: I've created ArchiveDeleted that I think that it solves this better

**Pull request merged**